### PR TITLE
[Op] Add Concat+Cast fusion

### PR DESCRIPTION
**Pull request by aalbersk**
Issue: https://github.com/changqi1/DeepRec/issues/24

### CODE REVIEW

**Comment by aalbersk** [comment]
@shanzhou2186 @changqi1 @Duyi-Wang  I ran the benchmark 3 times on AliCloud. You can see the results in the table below. The performance numbers shows the throughput (in MB/s) (higher is better).

| If Fused | Name                                           | Performance no. 1 \[MB/s\] | Performance no. 2 \[MB/s\] | Performance no. 3 \[MB/s\] | Average \[MB/s\] | Performance boost comparing to nonfused version |
| -------- | ---------------------------------------------- | -------------------------- | -------------------------- | -------------------------- | ---------------- | ----------------------------------------------- |
| fused    | concat\_cast\_2\_1\_fp32\_bfloat16\_512x16     | 770.00                     | 776.3                      | 771.2                      | 772.50           | 105%                                            |
| nonfused | concat\_cast\_2\_1\_fp32\_bfloat16\_512x16     | 738.1                      | 738.9                      | 727.1                      | 734.70           |
| fused    | concat\_cast\_13\_-1\_fp32\_bfloat16\_512x1    | 221.4                      | 223.5                      | 225.2                      | 223.37           | 120%                                            |
| nonfused | concat\_cast\_13\_-1\_fp32\_bfloat16\_512x1    | 184.6                      | 186.5                      | 185.3                      | 185.47           |
| fused    | concat\_cast\_13\_-1\_fp32\_bfloat16\_128x1    | 76.4                       | 75.6                       | 76                         | 76.00            | 112%                                            |
| nonfused | concat\_cast\_13\_-1\_fp32\_bfloat16\_128x1    | 68.1                       | 68                         | 67.9                       | 68.00            |
| fused    | concat\_cast\_16\_2\_fp32\_bfloat16\_512x32x16 | 3985.2                     | 3943.5                     | 3913.4                     | 3947.37          | 130%                                            |
| nonfused | concat\_cast\_16\_2\_fp32\_bfloat16\_512x32x16 | 3059                       | 3036.2                     | 3029.1                     | 3041.43          |
| fused    | concat\_cast\_64\_1\_bfloat16\_fp32\_400x60    | 3676.5                     | 3815.2                     | 3596.2                     | 3695.97          | 142%                                            |
| nonfused | concat\_cast\_64\_1\_bfloat16\_fp32\_400x60    | 2356.2                     | 2607.7                     | 2835                       | 2599.63          |

More data is available on Jira. Is there anything else you need?

File: tensorflow/core/kernels/concat_op.cc [hunk 1]
```diff
@@ -369,4 +370,209 @@ REGISTER_KERNEL_BUILDER(Name("ConcatOffset")
...
+      auto worker_threads = c->device()->tensorflow_cpu_worker_threads();
+      int num_threads = std::min(
+          static_cast<int>((output->NumElements() * sizeof(DstT)) / 4096),
+          worker_threads->num_threads);
+      if (num_threads == 1 || num_threads == 0) {
+        for (int64_t i = 0; i < inputs_flat_dim0; ++i) {
+          for (int64_t j = 0; j < N; ++j) {
+            auto size = sizes[j];
+            for (int64_t s = 0; s < size; ++s) {
+              out_ptr[s] = static_cast<DstT>(*input_pointers[j]++);
```
Comment (changqi1) [inline tensorflow/core/kernels/concat_op.cc:491]: @aalbersk  This is inefficient implementation. We could use SIMD code to convert the data type. like dnnl_cvt_float_to_bfloat16() and dnnl_cvt_bfloat16_to_float() API: https://github.com/alibaba/DeepRec/blob/main/third_party/mkl_dnn/oneDNN-v2.3.2-expose-bf16-tp.patch#L53-L55

Comment (changqi1) [reply tensorflow/core/kernels/concat_op.cc:491]: For BF16, oneDNN has given us a quick conversion API.

**Comment by changqi1** [comment]
@aalbersk  This is a real good job. Please tell us how to use grappler to find and replace concat+cast pattern and its kernel logic by meeting?

**Comment by changqi1** [comment]
> > @aalbersk This is a real good job. Please tell us how to use grappler to find and replace concat+cast pattern and its kernel logic by meeting?
> 
> Sure, I can explain it. How would you like to receive the explanation? Is a presentation enough?

Yes, a presentation is enough.

**Comment by aalbersk** [comment]
@changqi1 Shared presentation in Jira. Let me know if you need any more changes in the code.

**Comment by changqi1** [comment]
> @changqi1 Shared presentation in Jira. Let me know if you need any more changes in the code.

Good job.

**Comment by changqi1** [comment]
LGTM

**Pull request closed without merging**